File: ios/RNCalendar.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(RNCalendar, NSObject)

RCT_EXTERN_METHOD(addEvent:(NSString *)name startDate:(NSString *)starDate duration:(NSNumber *)duration description:(NSString *)description callback:(RCTResponseSenderBlock)callback)

@end

/*
//  RNCalendar.m
//  partiu
//
//  Created by diego.rios on 02/06/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import "RNCalendar.h"
#import <React/RCTLog.h>
#import <EventKit/EventKit.h>
#import <RCTConvert+RNSVG.h>

@implementation RNCalendar

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(addEvent:(NSString *)name location:(NSString *)location date:(NSString *)ISO8601DateString callback: (RCTResponseSenderBlock)callback)
{
  //RCTLogInfo(@"Pretending to create an event %@ at %@", name, location);
  NSString *message = [NSString stringWithFormat:@"%@:%@ at %@", @"Pretending to create an event", name, location];
  
  NSDate *date = [RCTConvert NSDate:ISO8601DateString];
  EKEventStore *store = [EKEventStore new];
  [store requestAccessToEntityType:EKEntityTypeEvent completion:^(BOOL granted, NSError * error) {
    if (!granted) { return; }
    EKEvent *event = [EKEvent eventWithEventStore:store];
    event.title = name;
    event.startDate = date;
    event.endDate = [event.startDate dateByAddingTimeInterval:60*60];
    event.calendar = [store defaultCalendarForNewEvents];
    NSError *err = nil;
    [store saveEvent:event span:EKSpanThisEvent commit:YES error:&err];
  }];
  
  callback(@[message]);
}

@end
*/

File: ios/RNCalendar.swift
//
//  RNCalendar.swift
//  partiu
//
//  Created by diego.rios on 03/06/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation
import EventKit

@objc(RNCalendar)
class RNCalendar: NSObject {
  
  @objc(addEvent:startDate:duration:description:callback:)
  func addEvent(name: String, startDate: NSNumber, duration: Double, description: String, callback: RCTResponseSenderBlock) -> Void {
    let eventStore = EKEventStore()
    let timeInterval = Double(truncating: startDate)/1000
    let date = Date(timeIntervalSince1970: timeInterval)
    
    switch EKEventStore.authorizationStatus(for: .event) {
      case .authorized:
        let response = insertEvent(
          store: eventStore,
          title: name,
          startDate: date,
          duration: duration,
          description: description
        )
        callback([response])
      case .denied:
        callback(["Access denied"])
      case .notDetermined:
        eventStore.requestAccess(to: .event, completion:
          {[weak self] (granted: Bool, error: Error?) -> Void in
            if granted {
              _ = self!.insertEvent(
                store: eventStore,
                title: name,
                startDate: date,
                duration: duration,
                description: description
              )
            } else {
              print("Access denied")
            }
          }
        )
        default:
          print("default")
      }
  }

  @objc
  func constantsToExport() -> [String: Any]! {
    return ["someKey": "someValue"]
  }
  
  func insertEvent(store: EKEventStore, title: String, startDate: Date, duration: Double, description: String) -> String {
    let event:EKEvent = EKEvent(eventStore: store)
    let _startDate = startDate
    let endDate = _startDate.addingTimeInterval(duration * 60 * 60)
    event.title = title
    event.startDate = _startDate
    event.endDate = endDate
    event.notes = description
    event.calendar = store.defaultCalendarForNewEvents
    do {
      try store.save(event, span: .thisEvent)
    } catch let error as NSError {
      return "Failed to save event with error: \(error)"
    }
    
    return "Event Saved"
    
  }

}
